go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,275 @@
+				// kube will restart and this will become a follower.
+				klog.Infof("No longer leader; exiting")
+				os.Exit(0)
+			},
+			OnNewLeader: func(newLeaderName string) {
+				if newLeaderName != cm.identity {
+					klog.Infof("Lost the election to %s; in standby mode", newLeaderName)
+				}
+			},
+		},
+	}
+
+	leaderelection.SetProvider(ovnkubeClusterManagerLeaderMetricsProvider{})
+	leaderElector, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	cm.wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		cm.wg.Done()
+	}()
+
+	return nil
+}
+
+// Run starts the watch factory, inits and runs the default network cluster controller
+func (cm *ClusterManager) Run() error {
+	metrics.RegisterClusterManagerFunctional()
+	if err := cm.defaultNetClusterController.Init(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		if err := cm.secondaryNetClusterControllerManager.Start(); err != nil {
+			return err
+		}
+	}
+
+	// Start and sync the watch factory to begin listening for events
+	if err := cm.wf.Start(); err != nil {
+		return err
+	}
+
+	if err := cm.defaultNetClusterController.Run(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		klog.Infof("Starting multi network attach manager")
+		return cm.secondaryNetClusterControllerManager.Run(cm.stopChan)
+	}
+
+	return nil
+}
+
+func (cm *ClusterManager) Stop() {
+	close(cm.stopChan)
+	cm.defaultNetClusterController.Stop()
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		cm.secondaryNetClusterControllerManager.Stop()
+	}
+}
+
+// secondaryNetworkClusterControllerManager object manages the multi net-attach-def controllers.
+// It implements networkAttachDefController.NetworkControllerManager and can be used
+// by NetAttachDefinitionController to add and delete NADs.
+type secondaryNetworkClusterControllerManager struct {
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *nad.NetAttachDefinitionController
+	client        clientset.Interface
+	ovnClient     *util.OVNClusterManagerClientset
+	kube          kube.Interface
+	watchFactory  *factory.WatchFactory
+}
+
+func newSecondaryNetworkClusterControllerManager(ovnClient *util.OVNClusterManagerClientset,
+	wf *factory.WatchFactory, recorder record.EventRecorder) *secondaryNetworkClusterControllerManager {
+	klog.Infof("Creating new multi network cluster manager")
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	sncm := &secondaryNetworkClusterControllerManager{
+		ovnClient:    ovnClient,
+		client:       ovnClient.KubeClient,
+		kube:         kube,
+		watchFactory: wf,
+	}
+	sncm.nadController = nad.NewNetAttachDefinitionController(
+		sncm, ovnClient.NetworkAttchDefClient, recorder)
+	return sncm
+}
+
+func (sncm *secondaryNetworkClusterControllerManager) Run(stopChan <-chan struct{}) error {
```

Comment:
Please see the comment above ?  I tried to do the way its done with the existing network controller manager and nadController.

Reply:
Like I mentioned above, it is not the same pattern.

So what we tried to follow is for a single receiver either Start/Stop or Init/Run (and not to confuse with the fact that some methods are exported for ease of testing).

So no mixing of Run/Stop or Init/Stop.

We had those different patterns before multi-net already and all this comes from trying to avoid to come up with more different patterns.

Reply:
Got it.  Your comment makes sense to me now.  Thanks for the patience.